alicloud/data_source_alicloud_vpcs.go

```diff
@@ -0,0 +1,198 @@
+				Optional: true,
+				ForceNew: true,
+			},
+			"status": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"name_regex": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validateNameRegex,
+				ForceNew:     true,
+			},
+			"is_default": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				ForceNew: true,
+			},
+			"vswitch_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			// Computed values
+			"vpcs": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"region_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"status": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vpc_name": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vswitch_ids": {
+							Type:     schema.TypeList,
+							Computed: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"cidr_block": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"vrouter_id": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"description": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"is_default": {
+							Type:     schema.TypeBool,
+							Computed: true,
+						},
+						"creation_time": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+func dataSourceAlicloudVpcsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AliyunClient).ecsconn
+
+	nameRegex, nameRegexOk := d.GetOk("name_regex")
+
+	args := &ecs.DescribeVpcsArgs{
+		RegionId: getRegion(d, meta),
+	}
+
+	var allVpcs []ecs.VpcSetType
+
+	for {
+		vpcs, paginationResult, err := conn.DescribeVpcs(args)
+		if err != nil {
+			break
+		}
+
+		allVpcs = append(allVpcs, vpcs...)
+
+		pagination := paginationResult.NextPage()
+		if pagination == nil {
+			break
+		}
+
+		args.Pagination = *pagination
+	}
+
+	var filteredVpcsTemp []ecs.VpcSetType
+
+	for _, vpc := range allVpcs {
+		if cidrBlock, ok := d.GetOk("cidr_block"); ok && vpc.CidrBlock != cidrBlock.(string) {
+			continue
+		}
+
+		if status, ok := d.GetOk("status"); ok && string(vpc.Status) != status.(string) {
+			continue
+		}
+
+		if isDefault, ok := d.GetOk("is_default"); ok && vpc.IsDefault != isDefault.(bool) {
+			continue
+		}
+
+		if vswitchId, ok := d.GetOk("vswitch_id"); ok && !vpcVswitchIdListContains(vpc.VSwitchIds.VSwitchId, vswitchId.(string)) {
+			continue
+		}
+
+		filteredVpcsTemp = append(filteredVpcsTemp, vpc)
+	}
+
+	var filteredVpcs []ecs.VpcSetType
+	if nameRegexOk {
+		r := regexp.MustCompile(nameRegex.(string))
+		for _, vpc := range filteredVpcsTemp {
```

Comment:
Line 145-146 can be more simplified by the following code:
if nameRegex, ok := d.GetOk("name_regex"); ok {
    r := regexp.MustCompile(nameRegex.(string))
}

In addition, can you ensure the code 'regexp.MustCompile(nameRegex.(string))' will be executed successfully?